### ISSUE TITLE
多显卡系统的功能优化

### ISSUE THREAD

**Issue by NaughtDZ** [WalkingMeatAxolotl/AnimaLoraStudio#491]
我有个笔记本电脑，2080MQ，然后雷电外接了魔改16G的3070显卡
几个问题：
1.一个是studio启动时候控制台显示的是3070卡，但是进去网页上面资源监控哪里也就是内存显存那里，显示的还是8G.

[studio] torch 2.11.0+cu128（GPU: NVIDIA GeForce RTX 3070）
[studio] flash_attn 启用
2026-08-11 09:07:07.769 INFO  studio.services.runtime.onnxruntime: [onnx_setup] DLL 搜索路径已加入 torch/lib（onnxruntime-gpu CUDA dlopen 用）

2.能否网页内有个可以切换计算训练显卡的地方
3.https://github.com/pollockjj/ComfyUI-MultiGPU   Comfyui有个节点能把模型拆到多个显卡显存里，显存总是比内存交换快的，能否看一看？

另外呢，就是在这个情况下训练kera2失败
----------------------------------------------
2026-08-11 09:13:25,550 - INFO - 加载配置文件: D:\AI\AnimaLoraStudio-0.22.0\studio_data\projects\2-baihumiaomiao\versions\v1\config.yaml
2026-08-11 09:13:29,354 - INFO - sample_seed=0 → 训练用随机种子: 1373158607
2026-08-11 09:13:29,385 - INFO - 📊 训练监控状态文件: D:\AI\AnimaLoraStudio-0.22.0\studio_data\tasks\6\monitor\state.json
2026-08-11 09:13:29,385 - INFO - 模型代码路径: D:\AI\AnimaLoraStudio-0.22.0\modeling\anima
2026-08-11 09:13:29,393 - INFO - 检测到 fp8 底模：以 fp8_base 语义训练（权重常驻 fp8，前向逐层 dequant）
2026-08-11 09:13:29,393 - INFO - 文本缓存已开启：先加载 VAE/Qwen3-VL，缓存并释放 TE 后再加载 Transformer
2026-08-11 09:13:29,394 - INFO - 加载 VAE...
2026-08-11 09:13:30,229 - INFO - VAE 权重加载: remap=none, 匹配 194/194 (100.0%), missing=0, unexpected=0
2026-08-11 09:13:30,555 - INFO - VAE 加载完成
2026-08-11 09:13:30,647 - INFO - 加载文本编码器...
2026-08-11 09:13:33,314 - INFO - [onnx_setup] DLL 搜索路径已加入 torch/lib（onnxruntime-gpu CUDA dlopen 用）
2026-08-11 09:13:33,371 - INFO - JSON caption 模式已启用（分类 shuffle）
2026-08-11 09:13:33,380 - INFO -   文件夹 1_data: 77 张 × repeat 1 × 分辨率[1024] = 77 样本
2026-08-11 09:13:33,380 - INFO - 数据集: 77 张图 → 77 样本（含 repeat）(JSON: 0, TXT: 77)
2026-08-11 09:13:33,414 - INFO - 检查 VAE latent 缓存（训练集）...
2026-08-11 09:13:33,495 - INFO - 所有 77 张图像已缓存（训练集）
2026-08-11 09:13:35,901 - INFO - VAE roundtrip 自检已保存: samples/vae_roundtrip.png
2026-08-11 09:13:35,915 - INFO - [text-cache] 预缓存 77 张图片 caption + 2 条采样/负面 prompt（varlen）
2026-08-11 09:13:35,942 - INFO - [text-cache] caption sidecar 全部命中（77 条），跳过编码
2026-08-11 09:13:36,245 - INFO - 加载 Krea2 Qwen3-VL（comfy 单文件形态）：D:\AI\AnimaLoraStudio-0.22.0\models\text_encoders\qwen3vl-4b-fp8\qwen3vl_4b_fp8_scaled.safetensors
2026-08-11 09:13:39,527 - INFO - Krea2 fp8 推理：252 个 Linear 挂 dequant 前向
2026-08-11 09:13:39,527 - INFO - Qwen3-VL fp8_scaled：252 层挂 dequant 前向
2026-08-11 09:13:41,429 - INFO - 文本缓存完成且文本编码器已释放；继续加载 Transformer...
2026-08-11 09:13:41,474 - INFO - attention_backend=none，flash_attn / xformers 都不启用，走 PyTorch SDPA
2026-08-11 09:13:41,474 - INFO - 加载 Transformer...
2026-08-11 09:13:41,474 - INFO - block swap：末尾 24 层将常驻内存，不载入显存
----------------------------------
没错，输出到block swap网页直接就显示训练失败了，以下是训练参数
model_family: krea2
transformer_path: D:/AI/AnimaLoraStudio-0.22.0/models/diffusion_models/krea2-raw-fp8-scaled.safetensors
vae_path: D:/AI/AnimaLoraStudio-0.22.0/models/vae/qwen_image_vae.safetensors
text_encoder_path: D:/AI/AnimaLoraStudio-0.22.0/models/text_encoders/qwen3vl-4b-fp8
text_encoder_cache: true
data_dir: D:\AI\AnimaLoraStudio-0.22.0\studio_data\projects\2-baihumiaomiao\versions\v1\train
resolution:
- 1024
aspect_ratio_limit: 2.0
reg_data_dir: ''
reg_caption: null
reg_weight: 1.0
flip_augment: true
prefer_json: true
caption_comfy_encoding: true
cache_latents: true
vae_cache_batch_size: 0
blocks_to_swap: 24
cache_encode_tiled: false
lora_type: lokr
lora_rank: 64
lora_alpha: 64.0
lokr_factor: 8
lora_dora: false
lora_rs: false
lora_dropout: 0.0
lora_rank_dropout: 0.0
lora_module_dropout: 0.0
lora_reg_dims: null
epochs: 50
max_steps: 3500
batch_size: 1
grad_checkpoint: true
grad_accum: 4
learning_rate: 1.0
lr_scheduler: none
optimizer_type: prodigy_plus_schedulefree
ppsf_d_coef: 1.0
ppsf_prodigy_steps: 0
ppsf_beta1: 0.9
ppsf_beta2: 0.99
ppsf_split_groups: true
ppsf_split_groups_mean: false
ppsf_use_speed: false
ppsf_fused_back_pass: false
ppsf_use_stableadamw: true
weight_decay: 0.0
kv_trim: false
vae_tiling: auto
noise_enhancement_type: none
timestep_sampling: krea2_shift
timestep_shift: 3.0
timestep_mix_low_prob: 0.0
timestep_schedule_shift: 1.0
timestep_shift_resolution_aware: false
infonoise_enabled: false
loss_type: mse
loss_weighting: none
masked_loss: false
grad_clip_max_norm: 1.0
mixed_precision: bf16
attention_backend: none
num_workers: 0
output_dir: D:\AI\AnimaLoraStudio-0.22.0\studio_data\projects\2-baihumiaomiao\versions\v1\output
output_name: baihumiaomiao_v1
save_every_epochs: 10
save_every_steps: 0
save_state_every_epochs: 0
save_state_every_steps: 0
seed: 42
resume_lora: null
resume_state: null
sample_every: 0
sample_steps: 0
sample_infer_steps: 28
sample_cfg_scale: 4.5
sample_sampler_name: euler
sample_scheduler: simple
sample_width: 0
sample_height: 0
sample_seed: 0
sample_negative_prompt: ''
sample_prompt: newest, safe, 1girl, masterpiece, best quality
sample_prompts: []
eval_validation_enabled: false
----------------------------------

另外补充一下，在任务管理器中观察，整个过程两张显卡的GPU占用和VRAM占用都没变化
以下是nvidia-smi的输出
 C:\Users\16037>nvidia-smi
Tue Aug 11 09:17:13 2026
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 610.88                 KMD Version: 610.88        CUDA UMD Version: 13.3     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                  Driver-Model | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 2080 ...  WDDM  |   00000000:01:00.0  On |                  N/A |
| N/A   57C    P0             31W /   80W |    1147MiB /   8192MiB |      1%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA GeForce RTX 3070      WDDM  |   00000000:07:00.0 Off |                  N/A |
| 30%   43C    P0             42W /  198W |     174MiB /  16384MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A            4328    C+G   ...y\StartMenuExperienceHost.exe      N/A      |
|    0   N/A  N/A            9800    C+G   C:\Windows\explorer.exe               N/A      |
|    0   N/A  N/A           10340    C+G   ...enCommandCenterBackground.exe      N/A      |
|    0   N/A  N/A           10372    C+G   ...les\Tencent\Weixin\Weixin.exe      N/A      |
|    0   N/A  N/A           10508    C+G   ..._cw5n1h2txyewy\SearchHost.exe      N/A      |
|    0   N/A  N/A           14560    C+G   ...5n1h2txyewy\TextInputHost.exe      N/A      |
|    0   N/A  N/A           17104    C+G   C:\Windows\explorer.exe               N/A      |
|    0   N/A  N/A           17388    C+G   ...xyewy\ShellExperienceHost.exe      N/A      |
|    0   N/A  N/A           17812    C+G   ....0.4129.72\msedgewebview2.exe      N/A      |
|    0   N/A  N/A           21864    C+G   ...ram Files\Tencent\QQNT\QQ.exe      N/A      |
|    0   N/A  N/A           22104    C+G   ...yb3d8bbwe\WindowsTerminal.exe      N/A      |
|    0   N/A  N/A           23360    C+G   ...acted\runtime\WeChatAppEx.exe      N/A      |
|    0   N/A  N/A           24960    C+G   ...les\Tencent\Weixin\Weixin.exe      N/A      |
|    0   N/A  N/A           25532    C+G   ...m Files\Tencent\QQNT\QQEX.exe      N/A      |
|    0   N/A  N/A           28132    C+G   ...rEngine\BaiduNetdiskUnite.exe      N/A      |
|    0   N/A  N/A           29444    C+G   ...ffice\root\Office16\EXCEL.EXE      N/A      |
|    1   N/A  N/A           21864    C+G   ...ram Files\Tencent\QQNT\QQ.exe      N/A      |
+-----------------------------------------------------------------------------------------+

**Comment by WalkingMeatAxolotl** [comment]
这个的确没有考虑涅，我自己是单显卡，不是很好测试，要看看怎么修改

**Comment by NaughtDZ** [comment]
> 这个的确没有考虑涅，我自己是单显卡，不是很好测试，要看看怎么修改

切换显卡问题，目前我的解决方案是在studio.bat中加
set CUDA_VISIBLE_DEVICES=0
这么一行，网页端顶部监控的虽然是2080，但是实际显存使用已经跑在3070里而且训练跑起来了。
但是咧，77张图，1重复，18层卸载，32G DDR4 2666内存，1024分辨率，fp8模型，1个半小时跑了5个epco，训练速度相当感人，显卡坞肯定有性能损失，但是这个问题真不在显卡坞上。
可以看到的是，非常长的时间GPU使用率为0%，资源查看器可见内存占用100%，也就是绝大多数性能完全浪费在内存通过雷电3跟显卡换数据了。要是真的能跟comfyui整合两个显存就牛逼了。
反过来说，同样的配置，跑家里的5090，8层卸载，凌晨挂个机，第二天一醒就完成了。
别问我为啥捣鼓笔记本接显卡训练，公司的电可不用我交钱，嘿嘿。

**Comment by NaughtDZ** [comment]
此图为笔记本外接3070

<img width="1678" height="847" alt="Image" src="https://github.com/user-attachments/assets/21eda3f3-4c7b-45ae-a272-2648ac727970" />

此图为昨晚5090（原来一个多小时就跑完了吗...）

<img width="1128" height="644" alt="Image" src="https://github.com/user-attachments/assets/fd3866ba-b5db-42b2-9296-b0f7c8a433e8" />

<img width="1636" height="1147" alt="Image" src="https://github.com/user-attachments/assets/baf68f15-e65a-4e4f-9d0e-654482a28694" />

**Comment by WalkingMeatAxolotl** [comment]
感觉可以，我加一个环境设置来切换显卡

**Comment by WalkingMeatAxolotl** [comment]
多 gpu 后面来看看分布式训练涅。现在对 block swap 可能收益有限，dram 到 vram 的读取是和传播是并行的，一般 x 个 block 还没有计算完，下一轮的 block 的读取就已经完成了。

**Comment by WalkingMeatAxolotl** [comment]
切换卡的功能0.24 加上了
多卡训练要等后续的版本涅

**Comment by NaughtDZ** [comment]
> 切换卡的功能0.24 加上了 多卡训练要等后续的版本涅

感谢，切显卡很丝滑，显存占用也显示对了
